### PR TITLE
WIP: Feature/universal rendering

### PR DESCRIPTION
**Pull request by gorghoa**
todo:
- [x] merge develop
- [x] links in dangerouslySetInnerHtml are triggering page reload, not react routing

### CODE REVIEW

File: src/containers/Page/Page.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
+import React, { Component, PropTypes } from 'react';
```
Comment (gorghoa) [inline src/containers/Page/Page.js:1]: interesting part

File: src/containers/Page/containers/PageContainer.js [hunk 2]
```diff
@@ -1,26 +1,26 @@
-import { connect } from 'react-redux'
```
Comment (gorghoa) [inline src/containers/Page/containers/PageContainer.js:1]: interesting part

File: src/containers/Page/index.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
+import { connect } from 'react-redux';
```
Comment (gorghoa) [inline src/containers/Page/index.js:1]: interesting part

File: src/containers/Page/index.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+ const mapStateToProps = (state) => ({
+ counter: tripleCount(state)
+ })
+
+ Selectors can compute derived data, allowing Redux to store the minimal possible state.
+ Selectors are efficient. A selector is not recomputed unless one of its arguments change.
+ Selectors are composable. They can be used as input to other selectors.
+ https://github.com/reactjs/reselect    */
+
+Page.reduxAsyncConnect = [{
```
Comment (gorghoa) [inline src/containers/Page/index.js:41]: Used to resolve jsonld page before first rendering. mandatory for server rendering

File: src/helpers/DataClient.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import superagent from 'superagent';
```
Comment (gorghoa) [inline src/helpers/DataClient.js:1]: intertesting part. Handle the fetching of jsonld files from browser or server.

File: src/helpers/DataClient.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+// import config from '../config';
+
+function formatUrl(path) {
+  const adjustedPath = path[0] !== '/' ? '/' + path : path;
+  return adjustedPath;
+}
+
+export function getPage(path) {
+  return new Promise((resolve, reject) => {
+    const toto = require('../static' + path);
```
Comment (gorghoa) [inline src/helpers/DataClient.js:11]: wrong variable name.

File: src/redux/modules/page.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
+import { combineReducers } from 'redux';
```
Comment (gorghoa) [inline src/redux/modules/page.js:1]: interesting part

File: src/containers/Page/Page.js (hunk 1 above)
Comment (dunglas) [reply src/containers/Page/Page.js:1]: Can you rebase to import changes from my last commits?

File: src/containers/Page/Page.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+      return {
+        __html: wrap.html()
+      };
+    }
+
+    render() {
+      return (
+            <div>
+                <Link to="core/getting-started" >REACT LINK THAT WORKS</Link>
```
Comment (dunglas) [inline src/containers/Page/Page.js]: Using https://github.com/reactjs/react-magic should fix that.

File: src/redux/modules/page.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+  return (dispatch) => {
+    dispatch(requestPage(pageName));
+    let jsonldDoc = pageName;
+    if (jsonldDoc === '' || jsonldDoc.endsWith('/')) {
+      jsonldDoc = jsonldDoc + 'index';
+    }
+
+    return getPage(`/data/${jsonldDoc}.jsonld`)
+              .then(data => {
```
Comment (dunglas) [inline src/redux/modules/page.js:108]: CS

File: src/containers/Page/Page.js (hunk 8 above)
Comment (gorghoa) [reply src/containers/Page/Page.js]: will try

File: src/redux/modules/page.js (hunk 9 above)
Comment (gorghoa) [reply src/redux/modules/page.js:108]: could you be more explicit on the mater ? It’s about the unneeded `{}`?

File: src/containers/Page/Page.js (hunk 1 above)
Comment (gorghoa) [reply src/containers/Page/Page.js:1]: Sure thing.

**Comment by gorghoa** [comment]
Just a status update: I’m in the middle of rebasing, but I’m struggling with the loading of Prisms assets on the server side. I’ll commit when this is fixed ;)

**Comment by dunglas** [comment]
@gorghoa in fact Webpack loadq all assets (it converts the dynamic call to a regex). Maybe can we just import all files explicitly for now.

**Comment by gorghoa** [comment]
@dunglas, there is a subtle difference yet: browser sides, assets are loaded throught http request, but serverside, they’re first bundled by webpack-isomorphic-tools in a `webpack-assets` file, then extracted from that file (feel weird doesn’t it ?…).
I managed to make it works for jsonlds files, but, as for now, I didn’t managed to make it load prismsjs assets yet.

**Comment by dunglas** [comment]
But if we import using all files explicitly, it will not be optimized but it will work both client-side and server-side isn't?

**Comment by dunglas** [comment]
Related:
- https://github.com/PrismJS/prism/issues/1020
- https://github.com/valor-software/prismjs-loader

**Comment by gorghoa** [comment]
Sorry, didn’t saw your feedbacks earlier.

Update status:
- [x] prismjs js is now workingish with server side rendering: code blocks are tokenized, but css is not loaded… (I can’t deny that I’m a bit tempted to hard code them in the html head…). Works fine when js is activated.
- [x] local links target are handled and does not trigger full page refresh anymore (click event are captured and handled manually)

todo:
- [ ] AnchorJS is using some dom methods, hence, does not work server side
- [ ] find a way to import prismjs css during server side rendering.

@dunglas, I think that calavera is generating html entities for code blocks (`&gt;`, `&lt;` and so forth), that does not play well with prismjs:

image

**Comment by gorghoa** [comment]
As we said, we could merge this PR now and find a fix for the two remaining issues later (anchorjs and server side code highlighting tokenization)

**Comment by dunglas** [comment]
Thanks @gorghoa

**Pull request merged**